java/com_spotify_voyager_jni_Index.cpp

```diff
@@ -750,7 +750,27 @@ void Java_com_spotify_voyager_jni_Index_saveIndex__Ljava_io_OutputStream_2(
   }
 }
 
-// TODO: Add asBytes
+jbyteArray Java_com_spotify_voyager_jni_Index_asBytes(JNIEnv *env,
+                                                      jobject self) {
+  try {
+    // save Index to MemoryOutputStream
+    std::shared_ptr<Index> index = getHandle<Index>(env, self);
+    std::shared_ptr<MemoryOutputStream> outputStream =
+        std::make_shared<MemoryOutputStream>();
+    index->saveIndex(outputStream);
+    // Write value of index to byte array
+    std::string nativeIndexValue = outputStream->getValue();
+    jbyteArray byteArray = env->NewByteArray(nativeIndexValue.length());
```

Comment:
I think it is 3 times? `n` bytes for `outputStream`, `n` bytes for `nativeIndexValue`, then `n` bytes for `byteArray`. There is only a shared_ptr to the index. Correct me if I'm wrong.

It's possible to get rid of the `nativeIndexValue` intermediate value but below we would have to call `outputStream.getValue()` 3 times inline for the byte array allocation and copying. The way it's implemented we can't get the length of the `MemoryOutputStream` without first getting the str value. 

Alternatively it seems I can get the length by tracking the position indicator of the stream (`std::ostream::tellp`), I'd have to add a getter method in `MemoryOutputStream` since the stream object is private. What do you think?

Reply:
> Could we implement asBytes() in Java instead of C++ by passing a [ByteArrayOutputStream](https://docs.oracle.com/javase/8/docs/api/java/io/ByteArrayOutputStream.html) into saveIndex?

`saveIndex()` takes a pointer to an `OutputStream` which is a custom class in StreamUtils.h. I don't think we can pass a ByteArrayOutputStream.

Reply:
Was able to use ByteArrayOutputStream + JavaOutputStream